### PR TITLE
topsql: try to reduce Plan Not Available

### LINKED ISSUE #31548: topsql: try to reduce Plan Not Available
## Dev task

image

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve?

Issue Number: close #31548

### What is changed and how it works?

Remove `""` plan when there are only 1 valid plan.

image

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [x] Performance regression: Consumes more CPU
- [x] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

N/A

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by crazycs520** [comment]
@breeswish @mornyx PTAL

File: util/topsql/collector/cpu.go [hunk 1]
```diff
@@ -235,6 +235,12 @@ func (s *sqlStats) tune() {
 		s.plans[""] = s.total
 		return
 	}
+	if len(s.plans) == 1 {
+		for k := range s.plans {
+			s.plans[k] = s.total
+			return
+		}
+	}
```
Comment (mornyx) [inline util/topsql/collector/cpu.go:243]:
```suggestion
	if len(s.plans) == 1 {
		s.plans[0] = s.total
		return
	}
```

File: util/topsql/reporter/datamodel.go [hunk 2]
```diff
@@ -497,10 +497,46 @@ func (c *collecting) appendOthersStmtStatsItem(timestamp uint64, item stmtstats.
...
 
+// removeInValidPlanRecord remove "" plan if there are only 1 valid plan in the record.
+func (c *collecting) removeInValidPlanRecord() {
+	sql2PlansMap := make(map[string][][]byte, len(c.records)) // sql_digest => []plan_digest
+	for _, v := range c.records {
+		k := string(v.sqlDigest)
+		sql2PlansMap[k] = append(sql2PlansMap[k], v.planDigest)
+	}
+	for k, plans := range sql2PlansMap {
+		if len(plans) != 2 {
```
Comment (mornyx) [inline util/topsql/reporter/datamodel.go:508]: Is this reasonable? Do we have only two `Plans` for the same `SQL`?

File: util/topsql/collector/cpu.go (hunk 1 above)
Comment (crazycs520) [reply util/topsql/collector/cpu.go:243]: `s.plans` is a map, not slice.

File: util/topsql/reporter/datamodel.go (hunk 2 above)
Comment (crazycs520) [reply util/topsql/reporter/datamodel.go:508]: No, but We can only remove the `""` plan when there are only 1 valid plan, so the `len(plans) == 2` is a filter to filter the SQL which `len(plans) == 2` and next filter use to filter at least 1 plan is `""` plan.

**Comment by mornyx** [comment]
LGTM

File: util/topsql/collector/cpu.go (hunk 1 above)
Comment (zhongzc) [inline util/topsql/collector/cpu.go:243]: Every second tune?

Comment (crazycs520) [reply util/topsql/collector/cpu.go:243]: Yes, this can reduce the pressure on the `repoter` behind.

File: util/topsql/reporter/datamodel.go [hunk 3]
```diff
@@ -497,10 +497,46 @@ func (c *collecting) appendOthersStmtStatsItem(timestamp uint64, item stmtstats.
 	others.appendStmtStatsItem(timestamp, item)
 }
 
+// removeInValidPlanRecord remove "" plan if there are only 1 valid plan in the record.
+func (c *collecting) removeInValidPlanRecord() {
```
Comment (zhongzc) [inline util/topsql/reporter/datamodel.go:501]:
```suggestion
func compactPlans(records) records {
```

File: util/topsql/collector/main_test.go [hunk 4]
```diff
@@ -74,6 +74,20 @@ func TestPProfCPUProfile(t *testing.T) {
 	require.Equal(t, []byte("sql_digest value"), data[0].SQLDigest)
 }
 
+func Test_sqlStats_tune(t *testing.T) {
```
Comment (hawkingrei) [inline util/topsql/collector/main_test.go:77]: should be ```TestSqlStatsTune```.

File: util/topsql/reporter/datamodel_test.go [hunk 5]
```diff
@@ -470,3 +470,56 @@ func Test_encodeKey(t *testing.T) {
 	key := encodeKey(buf, []byte("S"), []byte("P"))
 	assert.Equal(t, "SP", key)
 }
+
+func Test_removeInValidPlanRecord(t *testing.T) {
```
Comment (hawkingrei) [inline util/topsql/reporter/datamodel_test.go:474]: ```TestRemoveInValidPlanRecord```

File: util/topsql/reporter/datamodel.go (hunk 3 above)
Comment (crazycs520) [reply util/topsql/reporter/datamodel.go:501]: `compact` sounds meanless to the user, he still needs to read the comment or code...

**Comment by crazycs520** [comment]
@breeswish PTAL

File: util/topsql/reporter/datamodel.go [hunk 6]
```diff
@@ -497,10 +497,46 @@ func (c *collecting) appendOthersStmtStatsItem(timestamp uint64, item stmtstats.
...
+	sql2PlansMap := make(map[string][][]byte, len(c.records)) // sql_digest => []plan_digest
```
Comment (zhongzc) [inline util/topsql/reporter/datamodel.go:502]: Is there a way to reuse this tmp map?

File: util/topsql/reporter/datamodel_test.go [hunk 7]
```diff
@@ -470,3 +470,56 @@ func Test_encodeKey(t *testing.T) {
...
+		plan string
+		tss  []uint64
+		cpus []uint32
+	}{
+		{"SQL-1", "PLAN-1", []uint64{1, 2, 3, 5}, []uint32{1, 1, 1, 1}},
+		{"SQL-1", "PLAN-2", []uint64{1, 2, 5, 6}, []uint32{1, 1, 1, 1}},
+		{"SQL-2", "PLAN-1", []uint64{1, 2, 3, 4, 5, 6}, []uint32{2, 2, 2, 1, 1, 1}},
+		{"SQL-3", "PLAN-1", []uint64{1, 2, 3, 4, 5, 6}, []uint32{1, 2, 2, 1, 1, 1}},
+	}
+	assert.Equal(t, len(result), len(c1.records))
```
Comment (hawkingrei) [inline util/topsql/reporter/datamodel_test.go:510]: Please use ```require``` to replace ```assert```.

File: util/topsql/reporter/datamodel.go (hunk 3 above)
Comment (breeswish) [inline util/topsql/reporter/datamodel.go:501]:
```suggestion
func (c *collecting) removeInvalidPlanRecord() {
```

**Review by breeswish — COMMENTED** [review]
This PR is only a workaround for the plan not available problem, right?

**Comment by crazycs520** [comment]
> This PR is only a workaround for the plan not available problem, right?

yes

File: util/topsql/reporter/datamodel_test.go [hunk 8]
```diff
@@ -470,3 +470,56 @@ func Test_encodeKey(t *testing.T) {
 	key := encodeKey(buf, []byte("S"), []byte("P"))
 	assert.Equal(t, "SP", key)
 }
+
+func TestRemoveInValidPlanRecord(t *testing.T) {
```
Comment (zhongzc) [inline util/topsql/reporter/datamodel_test.go:474]:
```suggestion
func TestRemoveInvalidPlanRecord(t *testing.T) {
```

File: util/topsql/reporter/datamodel.go [hunk 9]
```diff
@@ -497,10 +497,46 @@ func (c *collecting) appendOthersStmtStatsItem(timestamp uint64, item stmtstats.
 	others.appendStmtStatsItem(timestamp, item)
 }
 
+// removeInvalidPlanRecord remove "" plan if there are only 1 valid plan in the record.
```
Comment (zhongzc) [inline util/topsql/reporter/datamodel.go:500]:
```suggestion
// removeInvalidPlanRecord remove "" plan if there are only 1 valid plan in the record.
// Basically, it should be called once at the end of the collection, currently in `getReportRecords`.
```

**Review by zhongzc — APPROVED** [review]

**Review by tangenta — APPROVED** [review]

**Review by zhongzc — APPROVED** [review]

**Review by zhongzc — APPROVED** [review]

**Review by zhongzc — APPROVED** [review]

**Comment by crazycs520** [comment]
reverted https://github.com/pingcap/tidb/pull/31877

**Review by tangenta — APPROVED** [review]

**Review by zhongzc — APPROVED** [review]

**Comment by crazycs520** [comment]
@ti-chi-bot /merge

**Pull request merged**